### ISSUE TITLE
WHy use area in calculating the loss  fn : prim_to_pcl_loss

### ISSUE THREAD

**Issue by mahmoudEltaher** [paschalidoud/superquadric_parsing#17]
Dear @paschalidoud ,

Kindly i want to inquire why  use area in calculating the loss  fn : prim_to_pcl_loss

loss = torch.einsum("ij,ij,ij->", [torch.mean(D, -1), probs, area])

I tried to remove it and run model but the result is not good.
and  i wonder why use this in equation ,however it is not mentioned in the original paper.

Waiting for you kindly response.

**Comment by paschalidoud** [comment]
Hi,

Multiplying with the area or the volume is the simplest way to control the contribution of its primitive in the loss based on how big/small that primitive is. For the experiments presented in the paper we don't use this regularization because we empirically observed that it didn't affect performance. 

So also in your case, if you remove it, results should look similar to the ones presented in the paper.

Best,
Despoina

**Comment by mahmoudEltaher** [comment]
Actually i already train the model without area but i found the result is too far from the result when the model train with area.

[Table_NQ1D9PSHL_with_out_area_without_deformation_without_change_weights.zip](https://github.com/paschalidoud/superquadric_parsing/files/6701550/Table_NQ1D9PSHL_with_out_area_without_deformation_without_change_weights.zip)


train the model by Running a command :
python scripts/train_network.py ../../../ShapeNetCore.v2/03001627/ tmp/ --use_deformations --use_sq --lr 1e-4 --n_primitives 20 --architecture tulsiani --dataset_type shapenet_v2 --use_chamfer --run_on_gpu --batch_size 16

then I test the model
python scripts/forward_pass.py demo/04379243/ tmp/ --model_tag "1c123801cd052c833bd248747cb27ae2" --n_primitives 20 --weight_file tmp/Table_NQ1D9PSHL_with_out_area_without_deformation_without_change_weights/model_49999 --use_sq --dataset_type shapenet_v2

I change loss in function prim_to_pcl_loss

loss = torch.einsum("ij,ij->", [torch.max(D, -1)[0], probs])

and the image Result as following ?

![Table_NQ1D9PSHL_with_out_area_without_deformation_without_change_weights](https://user-images.githubusercontent.com/32056889/123091642-e5423080-d429-11eb-860a-362ddbb04b3c.PNG)

and with area the image Result as following ?

![Table_905CZIPRI_with_both_loss_with_out_deformation](https://user-images.githubusercontent.com/32056889/123092758-1f600200-d42b-11eb-97d4-bd63b04dee17.PNG)

and the input image 
![Table_input](https://user-images.githubusercontent.com/32056889/123093256-c2b11700-d42b-11eb-8536-760b96677a19.PNG)



So it seems that area ,have a significant effect on the model result 

Appreciating your collaboration.

**Comment by paschalidoud** [comment]
Hi @mahmoudEltaher,

From the training statistics and the screenshot that you shared it seems that the pointcloud to primitve loss is lower than the primitive to point cloud. I would recommend to experiment with the weights of these loss terms so that the two loss terms decrease together. I  have already explained this in one of your previous issues, so please have a look at my answer there.

Since our paper doesn't contain results with tables, I am not really sure with what you compare.  Tables are a quite difficult object class because they contain thin structures e.g. legs. As I have already mentioned in another issue that you have posted, the Chamfer loss is very prone to local minima,. The second screenshot you are posting from training without area regularization is a classic case of such a local minima, where the network predicts a single primitive for the two legs. How many times have you retrained your model and you noticed these results? Does it look reasonable for samples from the training set? How does it look in other test samples? Please have a look at my answers in one of your previous issues for more details. 

Moreover, I am also not sure whether you are training with images as input or with voxel grids. Obviously images are more challenging since the input to the network is only in 2D.

To sum up, as I mentioned before the area is simply a regularization that shouldn't affect performance since it only controls the contribution of its primitive in the loss. At least this is what I observed 3 years ago when I was actively working on this project. 

Best,
Despoina
